src/mmj/lang/Formula.java

```diff
@@ -508,6 +474,53 @@ public Sym setTyp(final Map<String, Sym> symTbl, final String typS)
     }
 
     /**
+     * Determine count of left (+1) and right (-1) parens.
+     * Ideally the value will be 0 (balanced).
+     * Note that if the value is ever negative, this returns -1 immediately -- because there's already an error.
+     * @return int parenthesis count
+     */
+    public int countBalancedParentheses() {
+        int balanceCount = 0;
+        for (Sym symbol : sym) {
+            String symbolId = symbol.getId();
+            if ("(".equals(symbolId)) {
```

Comment:
I'm not super happy about this approach, because it's hard-coded to the symbols `(` and `)`, which need not exist, and need not be balanced if they exist. If we are to support something like this, I would prefer to discover parens through an analysis pass during the grammar creation stage. (In particular, there are many "paren-like" symbol pairs in set.mm, including `( )` `[ ]` `{ }` `<. >.` `(. ).` `[. ].` `[_ ]_` `<< >>` and possibly others I've forgotten.)